session/manager_http.go

```diff
@@ -295,31 +294,50 @@ func (s *ManagerHTTP) DoesSessionSatisfy(r *http.Request, sess *Session, request
 			return nil
 		}
 	case config.HighestAvailableAAL:
-		i := sess.Identity
-		if i == nil {
-			i, err = s.r.IdentityPool().GetIdentity(ctx, sess.IdentityID, identity.ExpandCredentials)
+		if sess.Identity == nil {
+			sess.Identity, err = s.r.IdentityPool().GetIdentity(ctx, sess.IdentityID, identity.ExpandNothing)
 			if err != nil {
 				return err
 			}
-			sess.Identity = i
-		} else if len(i.Credentials) == 0 {
-			// If credentials are not expanded, we load them here.
-			if err := s.r.PrivilegedIdentityPool().HydrateIdentityAssociations(ctx, i, identity.ExpandCredentials); err != nil {
+		}
+
+		i := sess.Identity
+		available := i.AvailableAAL
+		if len(available) == 0 || available == identity.NoAuthenticatorAssuranceLevel {
+			// Available is 0 if the identity was created before the AAL feature was introduced, or if the identity
+			// was directly created in the persister and not the identity manager.
+			//
+			// aal0 indicates that the AAL state of the identity is probably unknown.
+			//
+			// In either case, we need to fetch the credentials from the database to determine the AAL.
+			if len(i.Credentials) == 0 {
+				// The identity was apparently fetched without credentials. Let's hydrate them.
+				if err := s.r.PrivilegedIdentityPool().HydrateIdentityAssociations(ctx, i, identity.ExpandCredentials); err != nil {
+					return err
+				}
+			}
+
+			available = identity.NoAuthenticatorAssuranceLevel
+			if firstCount, err := s.r.IdentityManager().CountActiveFirstFactorCredentials(ctx, i); err != nil {
+				return err
+			} else if firstCount > 0 {
+				available = identity.AuthenticatorAssuranceLevel1
+			}
+
+			if secondCount, err := s.r.IdentityManager().CountActiveMultiFactorCredentials(ctx, i); err != nil {
 				return err
+			} else if secondCount > 0 {
+				available = identity.AuthenticatorAssuranceLevel2
 			}
-		}
 
-		available := identity.NoAuthenticatorAssuranceLevel
-		if firstCount, err := s.r.IdentityManager().CountActiveFirstFactorCredentials(ctx, i); err != nil {
-			return err
-		} else if firstCount > 0 {
-			available = identity.AuthenticatorAssuranceLevel1
-		}
+			i.AvailableAAL = available
 
-		if secondCount, err := s.r.IdentityManager().CountActiveMultiFactorCredentials(ctx, i); err != nil {
-			return err
-		} else if secondCount > 0 {
-			available = identity.AuthenticatorAssuranceLevel2
+			// This is the migration strategy for identities that already exist.
+			if managerOpts.upsertAAL {
```

Comment:
During the "migration" we just recompute it for everyone. Users that don't log in during the migration period stay at AAL0 forever. I think a nullable column is the way to go instead.
```suggestion
			// This is the migration strategy for identities that already exist.
			if !i.AvailableAAL.Valid {
```

Reply:
Yup - it's now nullable. This parameter however is still here because we don't always want to perform an INSERT or UPDATE statement when this method is called, so there has to be a way to control it.

Reply:
OK, makes sense.